gobblin-api/src/main/java/gobblin/configuration/SourceState.java

```diff
@@ -99,7 +98,7 @@ public SourceState(State properties, SourceState previousSourceState, List<WorkU
    * @return {@link SourceState} of the previous job run
    */
   public SourceState getPreviousSourceState() {
-    return new ImmutableSourceState(this.previousSourceState);
+    return new ImmutableSourceState(this.previousSourceState.or(new SourceState()));
```

Comment:
It was actually doing pre-mature objection creation of the `SourceState` class even before the other members are properly initialized, which is not a good practice.